src/Middleware/AuthorizationMiddleware.php

```diff
@@ -1,57 +1,169 @@
 <?php
-namespace App\Middleware;
+/**
+ * CakePHP(tm) : Rapid Development Framework (https://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (https://cakefoundation.org)
+ * @link          https://cakephp.org CakePHP(tm) Project
+ * @since         1.0.0
+ * @license       https://opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Authorization\Middleware;
 
-use Authorization\BouncerInterface;
-use Cake\Authorization\Bouncer;
+use Authorization\AuthorizationServiceInterface;
+use Authorization\IdentityDecorator;
+use Authorization\IdentityInterface;
+use Cake\Core\HttpApplicationInterface;
+use Cake\Core\InstanceConfigTrait;
+use InvalidArgumentException;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
 use RuntimeException;
 
 /**
- * Authorization Middleware
+ * Authorization Middleware.
  *
- * Injects the bouncer object into the request object as attribute
+ * Injects the authorization service and decorated identity objects into the request object as attributes.
  */
 class AuthorizationMiddleware
 {
+    use InstanceConfigTrait;
+
+    /**
+     * Default config.
+     *
+     * - `identityDecorator`: Identity decorator class name or a callable.
+     *
+     * @var array
+     */
+    protected $_defaultConfig = [
+        'identityDecorator' => IdentityDecorator::class,
+        'identityAttribute' => 'identity'
+    ];
 
     /**
-     * Bouncer class or instance
+     * Authorization service or application instance.
      *
-     * @var string||\Cake\Authorization\BouncerInterface
+     * @var \Authorization\AuthorizationServiceInterface|\Cake\Core\HttpApplicationInterface
      */
-    protected $bouncer = Bouncer::class;
+    protected $subject;
 
     /**
-     * Constructor
+     * Constructor.
      *
-     * @param string|\Cake\Authorization\BouncerInterface
+     * @param \Authorization\AuthorizationServiceInterface|\Cake\Core\HttpApplicationInterface $subject Authorization service or application instance.
+     * @param array $config Config array.
+     * @throws InvalidArgumentException
      */
-    public function __construct($bouncer)
+    public function __construct($subject, array $config = [])
     {
-        $this->bouncer = $bouncer;
+        if (!$subject instanceof AuthorizationServiceInterface && !$subject instanceof HttpApplicationInterface) {
+            $expected = implode('` or `', [
+                AuthorizationServiceInterface::class,
+                HttpApplicationInterface::class
+            ]);
+            $type = is_object($subject) ? get_class($subject) : gettype($subject);
```

Comment:
Don't we have a function in cakephp/core for this?

Reply:
It’s in the `3.next` branch.

Reply:
boooo